src/v/config/property.h

```diff
@@ -953,5 +963,118 @@ class hidden_when_default_property : public property<T> {
         return this->value() == this->default_value();
     }
 };
+namespace detail {
+
+template<typename P>
+concept Property = requires() {
+    std::derived_from<P, base_property>;
+    typename P::underlying;
+};
+
+template<typename T>
+concept Array = !(std::is_same_v<std::decay_t<T>, ss::sstring>
+                  || std::is_same_v<std::decay_t<T>, std::string>)
+                && is_collection<T>;
+
+template<typename T>
+concept OptArray = reflection::is_std_optional<T>
+                   && Array<typename T::value_type>;
+
+} // namespace detail
+
+template<detail::Property P>
+class enterprise : public P {
```

Comment:
> slightly un-ergonomic to construct.
variadic template prevents type deduction on initializer lists, so containers and aggregates need to be explicit at the call site

If you were to redesign the `enterprise` to operate by composition inside the `property`, instead of mixin, you should have all the types being concrete in the constructors. I think...
The downside is that every property object would be a bit bigger, regardless of whether it is describing an enterprise feature or not.

Not sure if it worth the effort, though. I kinda like this design!